### ISSUE TITLE
Support Bitcoin Cash Address format

### ISSUE THREAD

**Issue by Danconnolly** [bitcoinj-cash/bitcoinj#9]
Add support for Bitcoin Cash Address Format.

Links:
* https://github.com/Bitcoin-UAHF/spec/blob/master/cashaddr.md
* https://lists.linuxfoundation.org/pipermail/bitcoin-ml/2017-November/000472.html

Thoughts on Backwards Compatibility - Do we need to preserve backwards compatibility? Do we have any users yet? Should we replace the default method of obtaining string representations of an address with the new format, or add a new method?

**Comment by Danconnolly** [comment]
Some other implementations: https://github.com/Bitcoin-UAHF/spec/issues/36

**Comment by rafa-js** [comment]
I thought I sent my comments yesterday about this, but it turns out I didn't 😢 

I think we can include a module to normalize the addresses in the following way:
- If the address is invalid it throws an exception.
- If the address matches the old format it converts it to the new format.
- If the address matches the new format just keep it as is.

Something like:

```
interface IAddressNormalizer {
    String getNormalizedAddress(String address) throws IllegalAddressException;
}
```

This is the way we can preserve backwards compatibility also providing an extension-friendly solution: if the addresses format changes again, we just need to put our hand on this module without impact the rest of the project.

Since I'm not sure about the impact of the new address format within the project (where this new module should be placed) I could start coding the module and later someone with deeper understanding of this topic can integrate it appropriately. 

But first of all I'd like to know you opinion/validation about the requirements I propose.

Regards,

**Comment by HashEngineering** [comment]
The class that handles addresses is `org.bitcoinj.core.Address`.

As far as I understand, this class stores the address as a public key hash and it renders the address into a string (currently Base58) when needed based on the network being used at that time (mainnet, testnet, regtest, etc).

The conventions of `Address` should be used in terms of throwing exceptions.  I would think only this class needs to be modified and the rest of bitcoinj-cash doesn't need to be the wiser concerning which address format is being used.

**Comment by Danconnolly** [comment]
I agree, its only the representation of the address that needs an update and the Address class would be the appropriate place to do this. I think that adding methods for the new format is preferable to maintain backwards compatibility.

**Comment by rafa-js** [comment]
I agree the rest of bitcoinj-cash doesn't need to be the wiser concerning which address format is being used. But the address format validation shouldn't be done within the `Address` class. The `Address` class should contain the address parameters and the minimal code to delegate this responsability to other class. 

I believe the structure of things like this, or for instance, `Transactions` running their own validation logic, are wrong design decisions. This classes should only be encapsulating parameters. They are overloaded.

We can't rewrite the project and undo this decisions, obviously. But we can do small improvements on the fly, getting ready for the future. If some changes are required now, it's probably it need to be changed again.

**Comment by HashEngineering** [comment]
`Address` is derived from `VersionedChecksummedBytes` which is the class that contains the version and the data (without a checksum).  `VersionedChecksummedBytes` can convert both ways between its own format and the Base58 format.  The version is network specific and also indicates which type of address it is for (P2PKH or P2SH).

The bitcoin cash version is something like this:  `bitcoincash:qpzry9x8gf2tvdw0s3jn54khce6mua7lcw20ayyn`.  The part before the `:` is the network.  After decoding the rest, the version is the first byte, followed by the bytes ending in a checksum.  In this format the version contains information about the address type (P2PKH or P2SH) as well as the size of the data that follows.

The versions currently used `VersionedChecksummedBytes` (such as 0 for P2PKH and 5 for P2SH) are not the same values with the Bech32 format.  The rest of the data in both formats appears to be the same for hash160 addresses.

This java code may be useful in converting from the string format to byte[] in Java:https://github.com/SamouraiDev/bech32/blob/master/src/main/java/com/samourai/wallet/segwit/Bech32Util.java

This will be a lot of work.  All of the unit tests are with Base58 format.  It's an interesting challenge.  I am wondering how I will handle this in the Bitcoin Cash Wallet app!

**Comment by HashEngineering** [comment]
I have done some work and created a `CashAddress ` class which is able to read a cash address and convert it to the internal format and then convert it back to the cash address format.

A test was created using the examples of new and old address formats found here: 
 https://github.com/Bitcoin-UAHF/spec/blob/master/cashaddr.md.  My code works as far as reading both formats and comparing their encoded hash values and versions.

At this point the `CashAddress` class is not integrated into bitcoinj in my implementation thus far.  I will do some more work to see how to do that.

Update:  converting from the cash address format to the internal format and back to the cash address format doesn't work, yet.

**Comment by rafa-js** [comment]
By the way, @atlas-shrugging and I were starting to work on this. We should start assigning issues to the people. We have a few resources (people) and if all of we work on the same thing, we are wasting our time.

@HashEngineering Can you show us this code so we can go on it?

**Comment by HashEngineering** [comment]
The code will temporarily be here:
https://github.com/HashEngineering/bitcoinj-cash/commit/a8c7e1dcbd7dadacb7feaa9fce3496149583d028

**Comment by HashEngineering** [comment]
I fixed the bug that prevent converting from the internal format of VersionChecksummedBytes to the cash address format.  Now all unit tests are passing.  It is quick and dirty translation from the C++ code of bitcoin-abc.  It needs polishing.

Additionally, I was thinking about a scheme that would use your idea @rseibane, except instead of normalizing an address string to the cash address format, it would "normalize" to `Address `object.
```
interface IAddressNormalizer {
    Address getNormalizedAddress(String address) throws IllegalAddressException;
}
```
My thinking is that there would also be an `AddressFactory` that would have a list of `IAddressNormalizer` implementations.  That would include Bitcoin normalizers for Base58 (bitcoin and possiblyt the bitpay version - though it has been discontinued) and also the new cash address format.  `AddressFactory` would then correct Address object from its own `getNormalizedAddress(String address)`.

Hopefully we can get this updated soon, I would like my app recognizing the new address format.

**Comment by HashEngineering** [comment]
https://cashaddr.bitcoincash.org/ - Online Converter

**Comment by HashEngineering** [comment]
@rseibane - Any updates from your end? I haven't done anything since I posted the above comments.

**Comment by rafa-js** [comment]
Working on the tests. I hope to get this done by the end of this week. This is the list of cases I'm trying to cover:
- Prefix.
  - Prefix doesn't match with the checksum.
  - Prefix matches with the checksum.
- Version byte.
  - Unknown version byte.
  - P2KH version byte doesn't match.
  - P2SH version byte doesn't match.
  - P2KH version byte matches.
  - P2SH version byte matches.
- Checksum.
  - Invalid checksum.
  - Valid checksum.
- Uppercase/lowercase.
  - All lowercase.
  - All uppercase.
  - Mixing upper and lower.

Feel free to add any other.

**Comment by HashEngineering** [comment]
Some other tests can include:

- Convert from legacy to cash format using addresses generated by other tools.
- Convert from cash to legacy format using addresses generated by other tools.

**Comment by rafa-js** [comment]
Here is my current proposal for the tests: https://github.com/rseibane/bitcoinj-cash/commit/7907c7556a3c33301b5df759d22b2c2ee37c9d9a

Current case coverage:
- [X] Prefix.
  - [X] Prefix doesn't match with the checksum.
  - [X] Prefix matches with the checksum.
- [ ] Version byte.
  - [ ] Unknown version byte.
  - [ ] P2KH version byte doesn't match.
  - [ ] P2SH version byte doesn't match.
  - [ ] P2KH version byte matches.
  - [ ] P2SH version byte matches.
- [X] Checksum.
  - [x] Invalid checksum.
  - [X] Valid checksum. (This case is implicit in most of the tests)
- [X] Uppercase/lowercase.
  - [x] All lowercase.
  - [X] All uppercase.
  - [x] Mixing upper and lower.
- [X] Others.
  - [X] Convert from legacy to cash format using addresses generated by other tools.
  - [X] Convert from cash to legacy format using addresses generated by other tools.


Additional comments:
- The current `cashAddressCompareToLegacy` test has been splitted into smaller tests: `testFromLegacyToCashAddress` and `testFromCashToLegacyAddress`. The `cashAddressCompareToLegacy` has been deleted.
- The current `cashAddressIsValid` test I get the impression is kind of "test that the address conversion works in general terms". So this test has been divided into several smaller tests trying specific scenarios. The `cashAddressIsValid` has been deleted.
- I'm stuck with how to create addresses that only have some bytes wrong (version byte cases). If I modify just a byte the Checksum doesn't pass. I can't mock the checksum and force it to pass since this dependency is a static method. Please, hope this example helps you understand why we should avoid static methods. So: 
  - Do you know how to mock a static method using EasyMock?
  - Can you figure out how to create addresses whose byte checksum is right and some version bytes are wrong?
  - What about convert the `CashAddressHelper` static methods into instance methods? This way we can mock the verify checksum to true and just create addresses with wrong version bytes.

Regards,

**Comment by HashEngineering** [comment]
@rseibane 

The answer to create addresses with some bytes being incorrect is lies here:
```
    public String toString()
    {
        return CashAddressHelper.encode(getParameters().getCashAddrPrefix(), CashAddressHelper.packAddrData(getHash160(), type.getValue()));
    }
```

`CashAddressHelper` does not check all of the rules that `CashAddress` requires, much like how `VersionedChecksummedBytes` doesn't check the rules that `Address` requires concerning valid address versions.  `CashAddressHelper.encode` will create a checksummed address based on the data provided.

Ideally we should find a way to get invalid cash addresses from external sources.  This way we can test that our code is working.  At this point, we could have some bugs creating addresses and verifying them.

Unfortunately, the C++ code doesn't give too many test cases:
https://github.com/Bitcoin-ABC/bitcoin-abc/blob/master/src/test/cashaddr_tests.cpp
https://github.com/Bitcoin-ABC/bitcoin-abc/blob/master/src/test/cashaddrenc_tests.cpp

This javascript version is equally lacking of invalid addresses:
https://github.com/bitcoincashjs/cashaddrjs/blob/master/test/cashaddr.js

Regarding changing my code for `CashAddress` and `CashAddressHelper`, I am very much in favor of it.  These classes were quickly translated from C+ (hense all the static methods) and they are by no means a finished version.  We should change or refactor those classes in a way that makes sense for bitcoinj and and a scheme that satisfies your statement here: 

> But the address format validation shouldn't be done within the Address class. The Address class should contain the address parameters and the minimal code to delegate this responsibility to other classes.

By using this approach, we can minimize the required changes in the rest of bitcoinj.

**Comment by rafa-js** [comment]
This is nice, create addresses with some bytes incorrect gives us flexibility. In my mind, the first step is complete a good stack of tests. So, we can run the tests each time we make a change in the code and assure we aren't inserting a bug while refactoring the code. Having the functional code working is a good starting point.

The test we are writing now are more like white-box tests, each test is aware of the login behind the method. This tests are after reach a 100% of code coverage for the creating CashAddress methods.
Find out a set of invalid cash addresses from external sources would be cool, and it would be more like black-box tests. Anyway, I think we are good, we are having even more tests than the Bitcoin-ABC.

I'm trying to complete the proposed tests this evening.

Regards,

**Comment by rafa-js** [comment]
https://github.com/rseibane/bitcoinj-cash/commit/062a3919db91be456435ece638196e1f3ebb9473
I've just add the version byte tests. I noticed some of the proposed tests doesn't make a lot of sense, exactly the case of checking that the hash comes from a script hash or a public key hash.
This is the final list of included tests:

- [X] Prefix.
  - [X] Prefix doesn't match with the checksum.
  - [X] Prefix matches with the checksum.
- [X] Version byte.
  - [X] Unknown version byte.
  - [X] More than allowed padding.
  - [X] Non zero padding.
  - [X] First bit of version byte is non zero.
  - [X] Hash size doesn't match with the actual hash.
- [X] Checksum.
  - [x] Invalid checksum.
  - [X] Valid checksum. (This case is implicit in most of the tests)
- [X] Uppercase/lowercase.
  - [x] All lowercase.
  - [X] All uppercase.
  - [x] Mixing upper and lower.
- [X] Others (using the list of https://github.com/bitcoincashorg/spec/blob/master/cashaddr.md#examples-of-address-translation).
  - [X] Convert from legacy to cash format using addresses generated by other tools.
  - [X] Convert from cash to legacy format using addresses generated by other tools.

I'm optimistic with the current state of this feature. Next steps: find a bigger list of external addresses to test address conversion + refactor?

Regards,

**Comment by HashEngineering** [comment]
Great news.  I look forward to the finished code.

**Comment by rafa-js** [comment]
I have some updates. Two commits:

- https://github.com/rseibane/bitcoinj-cash/commit/c463eb238df5140b3b62ab740a8a74bfaaaff8c6 for the CashAddress refactoring.
- https://github.com/rseibane/bitcoinj-cash/commit/2014618de667b14afd9f111cab86e7d4a04fc52d for the integration within the legacy Address.

Some additional comments:

- I'm not done with the CashAddressHelper but it's enough at the moment.
- From now the `CashAddressFactory` should be used to get `Address` instances. So, the static methods within the `Address` class has been annotated has deprecated since it internally uses the `CashAddressFactory` to keep the compatibility but use the last address version.
- A few `Address` tests fail with the integration because it tries to do some comparisons using `address.toString()` vs base58 address, what now actually means `cashAddress.toString()` what actually means `prefix:payload`... What do you think should be done with this tests?
- Some files look hardy edited but is just because have been formatted.

Looking forward to your feedback.

Regards,

**Comment by HashEngineering** [comment]
I like what you have done with the refactoring.

For `CashAddressHelper`, there  may be a good reason to keep it in a separate class as it does handle a wider spectrum of data than simply hash160 for Addresses.  Some of the notes in the Cash Address format articles and specs say that it could later be used for smart contracts.  There are also other Base58 data types which include Private Keys, extended private and public keys (xprv, xpub), but I haven't seen anything about changing formats for those, so we won't be concerned for now.

Regarding the Address tests that fail, one idea is to change `address.toString()` to `address.toBase58()` in the tests.  `VersionedChecksummedBytes.toBase58()` is available.

**Comment by shadders** [comment]
On 1/02/2018 5:48 PM, Hash Engineering Solutions wrote:
> Some of the notes in the Cash Address format articles and specs say 
> that it could later be used for smart contracts.

I would expect this might refer to a payment protocol address.  Some of 
the more esoteric transaction types don't have an easy way to use a 
script template the way p2pkh and p2sh do.  In those cases the recipient 
needs to pass the output script in it's entirety to the payer.  But 
that's probably a fair way into the future.

**Comment by rafa-js** [comment]
Yes, I totally agree with keep the `CashAddressHelper` separeted. I'm thinking more about features like the name (CashAddressHelper sounds very generic as many as `CashAddressUtils `could be) or the big amount of comments (maybe this could be part of the documentation instead of the code). How about `CashAddressEncoding`?

About the tests, I will do some changes to make it pass but keeping the logic alright. As far as I concern, this is the last task to get this minimally done. 

Any update about the external list to test conversions?

Regards,

**Comment by HashEngineering** [comment]
I haven't found any more lists of addresses from external sources (valid and invalid) to serve as test data for this issue.

**Comment by Danconnolly** [comment]
I cant find any either, I've checked Electron Cash, Bitcoin Unlimited, and Bitcoin ABC, they all include the standard 6.

I'm thinking of generating a random set, say 1000 of them, based on addresses found in the blockchain, with conversion verified by at least 2 different implementations.

**Comment by rafa-js** [comment]
I like your way @Danconnolly 
I'm starting to collect a list of BCH addresses in the legacy format from here: https://bitinfocharts.com/top-100-richest-bitcoin%20cash-addresses.html. Then I will use https://cashaddr.bitcoincash.org/ to get it in the new format and done.

Regards,

**Comment by rafa-js** [comment]
Here we go https://drive.google.com/file/d/1lflmfQIkfx-VXdiYQdcdLw5WZed7UKkP/view?usp=sharing

**Comment by rafa-js** [comment]
I have just added the batch of addresses to test: https://github.com/rseibane/bitcoinj-cash/commit/5aaae259b5dd5cdcf3470f3b23622371ac23ba1d

I'm thinking about the integration of the cash address format with the legacy format. I'm not sure we should force the `Address` methods to create addresses (`Address.fromBase58`, `Address.fromP2SHScript`, etc) return `CashAddresses`. 

After I did it, many tests have started to fail what makes me think it's not a good idea. Whoever wants to use the new `CashAddress`, can use the `CashAddressFactory` to create one and use it as a legacy `Address` but the rest could remain as is. 
Not sure about the impact of this change. If you think we should integrate it it's just a matter of redo all the failing tests.

Please folks, share your thoughts. 

Regards,

**Comment by HashEngineering** [comment]
@rseibane I am looking at your repo and I ran the tests, finding that many have failed due to Legacy != CashAddress in string format.

Going forward, I would imagine that we would decide that the default address format (string) will be the CashAddress format.

Given this, we can can change Address.toString to output CashAddress format, instead of Base58, which it inherits from its super class.  This will require changing toString() in many tests to toBase58() when an Address is compared to a base58 (legacy) string.

In DeterministicKeyChainTest, we have this in serializeOneAccount.
```
final Address address = Address.fromBase58(UnitTestParams.get(), "n2nHHRHs7TiZScTuVhZUkzZfTfVgGYwy6X");
assertEquals(address, key1.toAddress(UnitTestParams.get()));
```
Where we get this error:
```
java.lang.AssertionError: 
Expected :bchtest:qr55pwmw5vape68ltut8gzrqcjj8qvfw7sr9mamss3
Actual   :n2nHHRHs7TiZScTuVhZUkzZfTfVgGYwy6
```
In this case, a legacy address string is created by ECKey.toAddress(...).toString().  The above change of making Address.toString should fix these errors.  It should fix the failed test in WalletTest.  This will also eliminate the need to change every address in every test that is in the legacy format. 

However, when I changed Address.toString() to match the return value of CashAddress.toString, this was the error:
```
java.lang.AssertionError: expected: org.bitcoinj.core.CashAddress<bchtest:qr55pwmw5vape68ltut8gzrqcjj8qvfw7sr9mamss3> but was: org.bitcoinj.core.Address<bchtest:qr55pwmw5vape68ltut8gzrqcjj8qvfw7sr9mamss3>
Expected :org.bitcoinj.core.CashAddress<bchtest:qr55pwmw5vape68ltut8gzrqcjj8qvfw7sr9mamss3> 
Actual   :org.bitcoinj.core.Address<bchtest:qr55pwmw5vape68ltut8gzrqcjj8qvfw7sr9mamss3>
```
assertEquals fails because the two objects being compared are not of the same class.  Perhaps `Address` should be handling everything instead of returning `CashAddress` objects.  More thinking on this is required.

Fixing BitcoinURITest will require other types of changes.

I am also trying to consider how this new address format will be handled by other software that uses bitcoinj.  

Another thing is that you forked your repo from mine, which is not forked from the this repo.  I am working on changing that soon.  Before you push your changes, you will need to create your own fork that forks from this repo.

**Comment by HashEngineering** [comment]
@rseibane Any updates on this?  Anything we can help with.

With my Bitcoin Cash Android wallet app, I am getting support questions regarding invalid addresses now that some services are using the CashAddr format.

BlockChair, the block explorer is one such example:
https://blockchair.com/bitcoin-cash/transaction/5102378ee46b4d05432fcfe20ce7a135d0733a4fc5af10f7f80b23ae5c818bb5

**Comment by rafa-js** [comment]
Sorry, I've been crazy busy last week.

My intuition is we should keep the legacy Address as is and use CashAddress in the news places we want to use it. This is basically when we create new transactions or we want to display an address to the user.

Once we decided what to do with this integration issue I will fork from this repo to be able to send the code properly.

**Comment by HashEngineering** [comment]
@Danconnolly @shadders We should get feedback from the other developers.

If we change the default format in bitcoinj-cash, then it will change for all software that uses it.

If we do not change the default format, but leave it as legacy format, then software based on bitcoinj-cash will need to be modified in some way to handle both address formats.

Regarding my own app, I don't know exactly how each of these options will work and how much coding I will need to do.

**Comment by rafa-js** [comment]
> If we do not change the default format, but leave it as legacy format, then software based on bitcoinj-cash will need to be modified in some way to handle both address formats.

This is my call to integrate this. This is like if the new cash-address format was a new available feature. If the developers want to integrate it, they do. 
Moreover I think this option requires less coding. In fact, almost zero new code. Just instead of use `Address.fromBase58` to create addresses to perform the transactions, developers should use the `CashAddressFactory`.

**Comment by HashEngineering** [comment]
Then I await your fork.  Submit it for our review.  I will put it into my app, and use the new cash format for one function, Sending, and l will report back as to the results.

**Comment by rafa-js** [comment]
I have just sent the PR.

**Comment by Danconnolly** [comment]
implemented in PR #55
